Explorer/Assets/DCL/UI/GenericContextMenu/GenericContextMenuController.cs

```diff
@@ -0,0 +1,177 @@
+            position.y -= viewInstance!.ControlsContainer.rect.height / 2;
+
+            Vector3 newPosition = Vector3.zero;
+            float minNonOverlappingArea = float.MaxValue;
+            foreach (ContextMenuOpenDirection enumVal in Enum.GetValues(typeof(ContextMenuOpenDirection)))
+            {
+                Vector2 offsetByDirection = GetOffsetByDirection(enumVal, offsetFromTarget);
+                Vector3 currentPosition = position + new Vector3(offsetByDirection.x, offsetByDirection.y, 0);
+                float nonOverlappingArea = CalculateNonOverlappingArea(overlapRect ?? backgroundWorldRect, GetProjectedRect(currentPosition));
+                if (nonOverlappingArea < minNonOverlappingArea)
+                {
+                    newPosition = currentPosition;
+                    minNonOverlappingArea = nonOverlappingArea;
+                }
+            }
+
+            return newPosition;
+        }
+
+        private float CalculateNonOverlappingArea(Rect rect1, Rect rect2)
+        {
+            float area1 = rect1.width * rect1.height;
+            float area2 = rect2.width * rect2.height;
+
+            Rect intersection = Rect.MinMaxRect(
+                Mathf.Max(rect1.xMin, rect2.xMin),
+                Mathf.Max(rect1.yMin, rect2.yMin),
+                Mathf.Min(rect1.xMax, rect2.xMax),
+                Mathf.Min(rect1.yMax, rect2.yMax)
+            );
+
+            float intersectionArea = 0;
+
+            if (intersection is { width: > 0, height: > 0 })
+                intersectionArea = intersection.width * intersection.height;
+
+            return area1 + area2 - intersectionArea;
+        }
+
+        private Rect GetProjectedRect(Vector3 newPosition)
+        {
+            Vector3 originalPosition = viewInstance!.ControlsContainer.localPosition;
+            viewInstance!.ControlsContainer.localPosition = newPosition;
+            Rect rect = GetWorldRect(viewInstance!.ControlsContainer);
+            viewInstance!.ControlsContainer.localPosition = originalPosition;
+
+            return rect;
+        }
+
+        private static Rect GetWorldRect(RectTransform rectTransform)
+        {
+            var corners = new Vector3[4];
+            rectTransform.GetWorldCorners(corners);
+            Vector2 min = corners[0];
+            Vector2 max = corners[2];
+            Vector2 size = max - min;
+            return new Rect(min, size);
+        }
+
+        protected override void OnViewClose()
+        {
+            controlsPoolManager.ReleaseAllCurrentControls();
+            inputData.ActionOnHide?.Invoke();
+        }
+
+        private void TriggerContextMenuClose() => isClosing = true;
+
+        protected override UniTask WaitForCloseIntentAsync(CancellationToken ct) =>
+            inputData.CloseTask != null ? UniTask.WhenAny(UniTask.WaitUntil(() => isClosing, cancellationToken: ct), inputData.CloseTask.Value) : UniTask.WaitUntil(() => isClosing, cancellationToken: ct);
```

Comment:
You can replace `isClosing` and the corresponding callback with `viewInstance.BackgroundCloseButton.OnClickAsync()`